Index: libiwinfo-2022-08-19-0dad3e66/iwinfo_nl80211.c
===================================================================
--- libiwinfo-2022-08-19-0dad3e66.orig/iwinfo_nl80211.c
+++ libiwinfo-2022-08-19-0dad3e66/iwinfo_nl80211.c
@@ -3358,8 +3358,8 @@ static int nl80211_get_mbssid_support(co
 
 static int nl80211_hardware_id_from_fdt(struct iwinfo_hardware_id *id, const char *ifname)
 {
-	char *phy, compat[64], path[PATH_MAX];
-	int i;
+	char *phy, *compat_p, compat[64], path[PATH_MAX], compat_part[64];
+	int i, part_len;
 
 	/* Try to determine the phy name from the given interface */
 	phy = nl80211_ifname2phy(ifname);
@@ -3370,57 +3370,65 @@ static int nl80211_hardware_id_from_fdt(
 	if (nl80211_readstr(path, compat, sizeof(compat)) <= 0)
 		return -1;
 
-	if (!strcmp(compat, "qca,ar9130-wmac")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x0029;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x9130;
-	} else if (!strcmp(compat, "qca,ar9330-wmac")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x0030;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x9330;
-	} else if (!strcmp(compat, "qca,ar9340-wmac")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x0030;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x9340;
-	} else if (!strcmp(compat, "qca,qca9530-wmac")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x0033;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x9530;
-	} else if (!strcmp(compat, "qca,qca9550-wmac")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x0033;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x9550;
-	} else if (!strcmp(compat, "qca,qca9560-wmac")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x0033;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x9560;
-	} else if (!strcmp(compat, "qcom,ipq4019-wifi")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x003c;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x4019;
-	} else if (!strcmp(compat, "qcom,ipq5018-wifi")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x003c;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x5018;
-	} else if (!strcmp(compat, "qcom,qcn6122-wifi")) {
-		id->vendor_id = 0x168c;
-		id->device_id = 0x003c;
-		id->subsystem_vendor_id = 0x168c;
-		id->subsystem_device_id = 0x6122;
-	} else if (!strcmp(compat, "mediatek,mt7622-wmac")) {
-		id->vendor_id = 0x14c3;
-		id->device_id = 0x7622;
-		id->subsystem_vendor_id = 0x14c3;
-		id->subsystem_device_id = 0x7622;
+	compat_p = compat;
+
+	while (part_len = strlen(compat_p)) {
+		strcpy(compat_part, compat_p);
+		compat_p += part_len+1;
+
+		if (!strcmp(compat_part, "qca,ar9130-wmac")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x0029;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x9130;
+		} else if (!strcmp(compat_part, "qca,ar9330-wmac")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x0030;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x9330;
+		} else if (!strcmp(compat_part, "qca,ar9340-wmac")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x0030;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x9340;
+		} else if (!strcmp(compat_part, "qca,qca9530-wmac")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x0033;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x9530;
+		} else if (!strcmp(compat_part, "qca,qca9550-wmac")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x0033;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x9550;
+		} else if (!strcmp(compat_part, "qca,qca9560-wmac")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x0033;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x9560;
+		} else if (!strcmp(compat_part, "qcom,ipq4019-wifi")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x003c;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x4019;
+		} else if (!strcmp(compat_part, "qcom,ipq5018-wifi")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x003c;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x5018;
+		} else if (!strcmp(compat_part, "qcom,qcn6122-wifi")) {
+			id->vendor_id = 0x168c;
+			id->device_id = 0x003c;
+			id->subsystem_vendor_id = 0x168c;
+			id->subsystem_device_id = 0x6122;
+		} else if (!strcmp(compat_part, "mediatek,mt7622-wmac")) {
+			id->vendor_id = 0x14c3;
+			id->device_id = 0x7622;
+			id->subsystem_vendor_id = 0x14c3;
+			id->subsystem_device_id = 0x7622;
+		}
 	}
+
 	return (id->vendor_id && id->device_id) ? 0 : -1;
 }
 
